Add interface methods for permission check for system tags

PVince81 opened the pull request:
Instead of checking for admin perm, use interface method
canUserAssignTag and canUserSeeTag to check for permissions.
Allows for more flexible implementation.
- [x] TODO: fix FIXME, allow admin to change perms again (need $isAdmin injection again)
- [x] TODO: adjust DAV unit tests

@nickvergessen can you have a look whether this is going to the right direction ?

Commit:
Add interface methods for permission check

Instead of checking for admin perm, use interface method
canUserAssignTag and canUserSeeTag to check for permissions.
Allows for more flexible implementation.

nickvergessen commented on lib/private/SystemTag/SystemTagManager.php:
```diff
@@ -316,6 +336,58 @@ public function deleteTags($tagIds) {
+	 * {@inheritdoc}
+	 */
+	public function canUserAssignTag($tag, $userId) {
+		if (!$tag instanceof ISystemTag) {
```
I'd just not allow that

PVince81 replied:
Okay.

How about also expecting a `User` object instead of `$userId` so we don't need to resolve it ?
It's always the "pass id" vs "pass object" question.

nickvergessen replied:
consistency is king I'd say and then also add typehints.

nickvergessen commented:
You put all the logic into the dav endpoint, I guess that's somewhat okay, on the other hand it could also be in the mapper? But then we would need a way around that for "forced" usage.

- [ ] Verify autotagger (CLI) based tagging

PVince81 commented:
- [ ] TODO: make the canUser* methods accept only ISystemTag and IUser

PVince81 commented:
> You put all the logic into the dav endpoint, I guess that's somewhat okay

I think that's what we agreed to do for the initial implementation.
The alternative would be to have all checks inside the managers, which we decided against. I don't remember why.

PVince81 commented:
> But then we would need a way around that for "forced" usage.

Okay, I think that was the reason ^

I'll leave the checks in the DAV endpoints for now to avoid changing too much.

Commit:
Updated canUser* functions in SystemTagManager to accept objects

Commit:
Add systemtag_group table and get/set methods

Added systemtag to group mapping table.
Added methods in ISystemTagManager to get/set the group mappings.

Commit:
Add system tag assignability check with groups

Whenever a user is not an admin, a tag is visible but not
user-assignable, check whether the user is a member of the allowed
groups.

PVince81 commented:
Added group mapping table + API + tests.

Commit:
Bring back admin permissions to change system tag permissions

Commit:
Fix system tags DAV unit tests

Commit:
Act on effective system tag canAssign permission

Whenever the server returns true for the can-assign Webdav property of
a system tag, it means the current user is allowed to assign,
regardless of the value of user-assignable.

This commit brings the proper logic to the web UI to make it possible
for users to assign when they have the permission.

PVince81 commented:
I think this is code complete now, needs more testing.

PVince81 commented:
Hmmm, maybe one more:
- [ ] extend integration tests

PVince81 commented:
- [ ] endpoint extension to get/set groups

Commit:
Add admin-only system tag groups property

For setting/getting system tag groups

PVince81 commented:
- [ ] TODO: unit test for GROUPS attribute

Commit:
Fix system tag update code

Commit:
Fix system tags DAV plugin tests

PVince81 commented:
will fix the tests

Commit:
Adjust DAV SystemTagPlugin unit tests for groups

Commit:
Update integration tests for system tags with groups

PVince81 commented:
Integration tests added.

Please review @nickvergessen @rullzer @SergioBertolinSG @blizzz

nickvergessen commented on apps/dav/lib/SystemTag/SystemTagPlugin.php:
```diff
@@ -222,8 +238,27 @@ public function handleGetProperties(
+		});
+
+		$propFind->handle(self::GROUPS_PROPERTYNAME, function() use ($node) {
+			if (!$this->groupManager->isAdmin($this->userSession->getUser()->getUID())) {
```
`getUID` called on potential null?

PVince81 replied:
shouldn't happen as we don't register this plugin when the user isn't logged in or in public page mode

nickvergessen commented on apps/dav/lib/SystemTag/SystemTagPlugin.php:
```diff
@@ -250,22 +286,42 @@ public function handleUpdateProperties($path, PropPatch $propPatch) {
+			}
+
+			if (isset($props[self::GROUPS_PROPERTYNAME])) {
+				if (!$this->groupManager->isAdmin($this->userSession->getUser()->getUID())) {
```
same here

nickvergessen commented on lib/private/SystemTag/SystemTagManager.php:
```diff
@@ -316,7 +333,104 @@ public function deleteTags($tagIds) {
+	public function getTagGroups(ISystemTag $tag) {
+		$groupIds = [];
+		$query = $this->connection->getQueryBuilder();
+		$query->select('*')
```
`select('gid')` ?

nickvergessen commented on lib/public/UserNotFoundException.php:
```diff
@@ -0,0 +1,62 @@
+ *
+ * @since 9.1.0
+ */
+class UserNotFoundException extends \RuntimeException {
```
Move in User\ ?

PVince81 replied:
Could do, this assuming that we'll eventually move all the interfaces \OCP\IUser* to \OCP\User\IUser eventually ?

nickvergessen replied:
Namespacing is one of the longterm goals

PVince81 replied:
Turns out it's unused now, it was in my original implementation that was based on user id strings instead of user objects.

nickvergessen commented on lib/private/SystemTag/SystemTagManager.php:
```diff
@@ -316,7 +333,104 @@ public function deleteTags($tagIds) {
+			throw $e;
+		}
+
+		return false;
```
The interface has no documented return value, and since you throw on error, you can remove this.

PVince81 replied:
Seems copy-paste strength was a bit high lately

nickvergessen commented on lib/public/systemtag/isystemtagmanager.php:
```diff
@@ -113,4 +116,49 @@ public function updateTag($tagId, $newName, $userVisible, $userAssignable);
+	 * @param ISystemTag $tag tag to check permission for
+	 * @param IUser $user user to check permission for
+	 *
+	 * @return true if the user is allowed to assign/unassign the tag, false otherwise
```
no, this is about seeing

PVince81 replied:
fixed

SergioBertolinSG commented on build/integration/features/tags.feature:
```diff
@@ -21,6 +21,18 @@ Feature: tags
+    Given user "user0" exists
+    When "admin" creates a "not user-assignable" tag with name "TagWithGroups" and groups "group1|group2"
+    Then The response should have a status code "201"
+	And The "not user-assignable" tag with name "TagWithGroups" has the groups "group1|group2"
```
Indentation.

SergioBertolinSG commented on build/integration/features/tags.feature:
```diff
@@ -122,6 +147,23 @@ Feature: tags
+  Scenario: Assigning a not user-assignable tag to a file shared by someone else as regular user belongs to tag's groups should work
+    Given user "user0" exists
+    Given user "user1" exists
+	Given group "group1" exists
```
Indentation.

nickvergessen commented on apps/dav/lib/SystemTag/SystemTagPlugin.php:
```diff
@@ -181,14 +185,26 @@ private function createTag($data, $contentType = 'application/json') {
-			return $this->tagManager->createTag($tagName, $userVisible, $userAssignable);
+			$tag = $this->tagManager->createTag($tagName, $userVisible, $userAssignable);
+			if (!empty($groups)) {
+				$this->tagManager->setTagGroups($tag, $groups);
```
Exceptions thrown by this will break DAV?

PVince81 replied:
`setTagGroups` doesn't throw any known tag-related exceptions.
The ones that could happen are rare DB exceptions, in which case it gets converted to a SabreException at the top level.

SergioBertolinSG commented on build/integration/features/tags.feature:
```diff
@@ -45,6 +57,19 @@ Feature: tags
+    Given "admin" creates a "not user-assignable" tag with name "TagWithGroups" and groups "group1|group2"
+    When "admin" edits the tag with name "TagWithGroups" and sets its groups to "group1|group3"
+    Then The response should have a status code "207"
+	And The "not user-assignable" tag with name "TagWithGroups" has the groups "group1|group3"
```
Indentation.

SergioBertolinSG commented on build/integration/features/tags.feature:
```diff
@@ -368,3 +410,18 @@ Feature: tags
+
+  Scenario: User can assign tags when in the tag's groups
+    Given user "user0" exists
+	Given group "group1" exists
```
Indentation.

PVince81 replied:
are we using tabs or spaces here ? In PHP and JS we're using tabs everywhere. I guess that here it's spaces then ?

nickvergessen replied:
yes, because it's yaml based and yaml uses spaces because space-intendation has a meaning

PVince81 replied:
ah, I didn't realize this was yaml!

strangely the tests all ran fine in my env, so maybe the test runner doesn't care.

Will replace these to spaces then.

SergioBertolinSG commented on build/integration/features/tags.feature:
```diff
@@ -368,3 +410,18 @@ Feature: tags
+	Given user "user0" belongs to group "group1"
+    When "admin" creates a "not user-assignable" tag with name "TagWithGroups" and groups "group1|group2"
+    Then The response should have a status code "201"
+	And the user "user0" can assign the "not user-assignable" tag with name "TagWithGroups"
```
Indentation.

SergioBertolinSG commented on build/integration/features/tags.feature:
```diff
@@ -368,3 +410,18 @@ Feature: tags
+    Given user "user0" exists
+    When "admin" creates a "not user-assignable" tag with name "TagWithGroups" and groups "group1|group2"
+    Then The response should have a status code "201"
+	And the user "user0" cannot assign the "not user-assignable" tag with name "TagWithGroups"
```
Indentation.

Commit:
Add app to shipped.json

Commit:
Rename "not-assignble" to "restricted"

Commit:
Some tweaks for systemtagmanager

PVince81 commented:
Fixed most of the comments and also fixed the indentations.
Also renamed "not assigned" to "restricted" as discussed before.

Please have another look @nickvergessen @SergioBertolinSG

PVince81 commented:
Ah, I haven't made "admin" the default group yet. Still having doubts about this approach.

MTRichards commented:
@PVince81 say more on those doubts.  You are thinking that a default group should not be set, or...?

PVince81 commented:
@MTRichards 

If we make "admin" the default group it is fine for new usages. However for people upgrading from 9.0 we need to write a repair step / migration to automatically add "admin" for all "not assignable" tags, since they used to be restricted to admins. If we don't do this, admins will lose the ability to assign such tags since their group hasn't been explicitly added to the list.

The alternative to repair steps would be to have the implicit logic: if no group is specified, give permission to admins. If at least one group is given, restrict the permission to this group.
This isn't too nice UX-wise as it might be confusing for admins.

PVince81 commented:
Or the much easier alternative: just keep everything as it is now: regardless of permissions, admins can *always* assign tags.

SergioBertolinSG commented:
Works and indentation is fine :+1:

PVince81 commented:
@MTRichards if being able to remove the admin group's permission for assigning doesn't bring a clear and important benefit, I'd suggest to not add this possibility as it adds more code paths and complexity, and simply keep what we have now: admins can always do everything.

MTRichards commented:
I think it will be quite important to remove the admin group as an owner of tagged files, but perhaps as a first step we keep it simple since admins can also do other things with tags that would largely break this (like make every uploaded file get the tag automatically) or in other ways be malicious.

Later, we can look to isolate this more - but I am ok as a step one to do as you suggest @PVince81

PVince81 commented:
@MTRichards thanks, you're right about the other ways for admins of setting tags.

@nickvergessen second review ?

Commit:
Make it possible to exclude admin groups from groups select2

nickvergessen commented:
- [ ] missing version.php increment, so the table did not get created for me

nickvergessen commented:
* When I open the page as user of a group with permissions, then remove the user from the group and then as a user (un)assign a reserved tag, that "seems to work", at least in the UI the state is changed, but after refreshing the changes are not done. So seems like the error from the failed assigning is not displayed.
However since that is only a race condition with very little use of real world, I think it doesn't matter too much?

* Can not create a new tag with the name of a reserved tag that is visible. Low priority for me as well, you can even work around by naming it differently and then rename the tag.

So :+1: when the version.php is fixed.

Commit:
Increase version for DB addition

PVince81 commented:
> However since that is only a race condition with very little use of real world, I think it doesn't matter too much?

Yeah, I'd rather fix this as bugfix. This could also happen if the tag got deleted by someone else concurrently.

Thanks for the reviews ! I added version.php

PVince81 commented:
Rebased
